src/Network/PeerDiscovery/Operations.hs

```diff
@@ -59,12 +111,35 @@ bootstrap pd node = join . atomically $ readTVar (pdBootstrapState pd) >>= \case
           targetId <- randomPeerId
           -- Populate part of the routing table holding nodes far from us.
           if testPeerIdBit myId 0 /= testPeerIdBit targetId 0
-            then peerLookup pd targetId >> result BootstrapDone
+            then do
+                   _ <- peerLookup pd targetId
+                   return ()
             else loop
-  where
-    result st = do
-      atomically $ writeTVar (pdBootstrapState pd) BootstrapNeeded
-      return $ st == BootstrapDone
+        putMVar initialPingMV True
```

Comment:
Check the exception handling here. If we get an exception in the response handler I presume the other handler for `sendRequest` does not get run. In which case here the `putMVar initialPingMV` would never happen and the main thread would hang in reading that MVar (or if it's lucky get a deadlocked on MVar exception).

We might want to add another util like the existing `sendRequestSync`, to use more symmetrically (and avoiding exception issues) something like:

```
newtype Async a = Async (MVar a)

sendRequestAsync
  :: (IsRequest req, Show req)
  => PeerDiscovery cm
  -> req                        -- ^ Request to be sent
  -> Node                       -- ^ Recipient
  -> IO (Async (ResponseType req))
sendRequestSync pd reqType peer = do
  v <- newEmptyMVar
  sendRequest pd reqType peer
    (putMVar v Nothing)
    (putMVar v . Just)
  return (Async v)

awaitAsync (Async v) = readMVar v
```
And then used like
```
  -- Send both ping requests. Though actually sending the second is conditional.
  pingAsync1 <- sendRequestAsync pd (Ping Nothing)
  pingAsync2 <- maybe (return Nothing)
                      (\port -> Just <$> sendRequestAsync pd (Ping (Just port)))
                      requestedPublicPort
  -- Wait for both responses (or failure). Again the second is conditional.
  ping1Res <- awaitAsync pingAsync1
  ping2Res <- maybe (return Nothing) (fmap Just . awaitAsync) pingAsync2

  case ping1Res of
    Nothing -> ...
    Just Pong -> ...

  case ping2Res of
    Nothing -> ... -- was not needed
    Just Nothing -> ... -- failure
    Just (Just Pong) -> ...
```